### ISSUE TITLE
Navigation over long distances

### ISSUE THREAD

**Issue by Subhashini2610** [mapbox/mapbox-navigation-ios#3778]
### Mapbox Navigation SDK version

2.0.1

### Steps to reproduce

I need to enable turn-by-turn navigation that uses a list of silent waypoints coming from my custom server. This navigation typically spans over tens to hundreds of kilometres (distance between the start and destination). Hence, the number of waypoints coming from the server is also huge (typically 100-200 waypoints) for the route to be determined. Please note that these silent waypoints only determine the route and are not stops for the user. We cannot afford any deviation from the route suggested by our custom server.

The problem is that the **NavigationRouteOptions** can take a max of 25 waypoints. Anything beyond and I get the following error:
```
The operation couldn't be completed. Too many coordinates; maximum number of coordinates is 25
```
Below is my code:
```
/// custom server
NetworkManager.shared.getRoute(startLon: 9.1407, startLat: 48.6980, destLon: 9.0074, destLat: 48.6845) { [unowned self] routeResponse, error in
    if (error != nil) {
        debugPrint("Error in fetching route: \(String(describing: error?.description))")
    } else {
        debugPrint("Route received")
        if let route = routeResponse?.route {
            
            // waypoints from server
            var lineCoordinates = [Waypoint]()
            
            for location in route {
                let waypoint = Waypoint(coordinate: CLLocationCoordinate2DMake(location.lat, location.lon))
                waypoint.separatesLegs = false
                lineCoordinates.append(waypoint)
            }
            
            //lineCoordinates.count is 180
            
            let options = NavigationRouteOptions(waypoints: lineCoordinates)
            
            Directions.shared.calculate(options) { [unowned self] (_, result) in
                switch result {
                case .failure(let error):
                    print(error.localizedDescription)
                case .success(let response):
                    
                    //  For demonstration purposes, simulate locations if the Simulate Navigation option is on.
                    //  Since first route is retrieved from response `routeIndex` is set to 0.
                    let navigationService = MapboxNavigationService(routeResponse: response, routeIndex: 0, routeOptions: options, simulating: isSimulationEnabled ? .always : .never)
                    let navigationOptions = NavigationOptions(navigationService: navigationService)
                    let navigationViewController = NavigationViewController(for: response, routeIndex: 0, routeOptions: options, navigationOptions: navigationOptions)
                    navigationViewController.modalPresentationStyle = .fullScreen
                    // Render part of the route that has been traversed with full transparency, to give the illusion of a disappearing route.
                    navigationViewController.routeLineTracksTraversal = true
                    navigationViewController.delegate = self
                    
                    present(navigationViewController, animated: true, completion: nil)
                }
            }
            
        }
    }
}
```
@andreas-ullrich and I are working together on this piece.

### Expected behavior

The navigation with silent waypoints should work seamlessly over larger distances as well.

### Actual behavior

This is the response of _**Directions.shared.calculate(options)**_

<img width="588" alt="Screenshot 2022-03-10 at 10 56 52 AM" src="https://user-images.githubusercontent.com/54573581/157595546-a7e1db56-b16d-4531-a2bf-8f68b70629c8.png">

### Is this a one-time issue or a repeatable issue?

repeatable

**Comment by Udumft** [comment]
Hello @Subhashini2610! 
25 waypoints is the limit of our Directions endpoint (please see this [doc section](https://docs.mapbox.com/api/navigation/directions/#directions-api-restrictions-and-limits)). To extend that, you can [contact](https://www.mapbox.com/contact/sales/) sales team to discuss your use-case and needs.

Other than that, there are other things you can try.
25 waypoints does not limit the distance of the route, but since you are mapping each route coordinate into a waypoint - than yes, 25 points limit will be exceeded quite quickly. First of all, I'd suggest updating to latest NavSDK version (currently [v2.8.1](https://github.com/mapbox/mapbox-navigation-ios/releases/tag/v2.8.1)) which will give you access to [RoutingProvider](https://github.com/mapbox/mapbox-navigation-ios/blob/v2.8.1/Sources/MapboxCoreNavigation/RoutingProvider.swift#L9) which is designed for user-implemented mechanisms for route calculations. Then all of the related logic will be encapsulated inside and SDK will handle it automatically for route building, rerouting, refreshing and etc. You can find an example on how to provide custom implementation [here](https://github.com/mapbox/mapbox-navigation-ios-examples/blob/main/Navigation-Examples/Examples/Custom-RoutingProvider.swift). Then, to build the route you need you can try 3 options:
1. Split your `routeResponse?.route` into 25 points segments and do individual requests to `Directions` endpoint. Then, combine the resulting routes in one final route, by manually creating [`Route`](https://github.com/mapbox/mapbox-directions-swift/blob/main/Sources/MapboxDirections/Route.swift#L23) object combining it's shape, legs, and other properties. Note then that it will increase the number of **billed** requests done to the endpoint!
2. Create `Route` object yourself directly with your `routeResponse?.route` coordinates. Transforming coordinates into route shape is a straightforward action but in this case you are responsible of acquiring congestion, visual and voice instructions and other annotations yourself.
3. Spread these 25 waypoints evenly across the entire route to make it follow the desired shape. This method depends on the complexity of your shape and 25 'checkpoints' may be not sufficient in all cases. It is also possible to apply more sophisticated heuristics to correctly place them. For example your custom server may provide you hints on best spots to do so, or you could start requesting with just Origin and Destination waypoints and compare your and Directions routes step-by-step iteratively to put next silent waypoint whenever deviation point occurs, then re-requesting it again and thus covering the entire route. Depending on the algorithm it may also increase amount of **billed route requests**!

To sum up: 25 waypoints is a software limit you can discuss with our support/sales team. But there are still potential options to work around that using custom routing logic in your app.
I am going to close this one since this is not a bug on your side, but please feel free to comment if you need any more information on that topic.
